File: Mixter.Domain.Tests/Identity/SessionTest.cs
using System;
using Mixter.Domain.Identity;
using Mixter.Domain.Identity.Events;
using NFluent;
using Xunit;

namespace Mixter.Domain.Tests.Identity
{
    public class SessionTest
    {
        private static readonly SessionId SessionId = SessionId.Generate();
        private static readonly UserId UserId = new UserId("user@mixit.fr");

        private readonly EventPublisherFake _eventPublisher = new EventPublisherFake();

        [Fact]
        public void WhenUserLogoutThenRaiseUserDisconnected()
        {
            var session = new Session(new UserConnected(SessionId, UserId, DateTime.Now));

            session.Logout(_eventPublisher);

            Check.That(_eventPublisher.Events)
                 .Contains(new UserDisconnected(SessionId, UserId));
        }
    }
}

File: domain/src/test/java/mixter/domain/identity/SessionTest.java
package mixter.domain.identity;

import mixter.domain.DomainTest;
import mixter.domain.Event;
import mixter.domain.SpyEventPublisher;
import mixter.domain.identity.events.UserConnected;
import mixter.domain.identity.events.UserDisconnected;
import org.junit.Before;
import org.junit.Test;

import java.time.Instant;

import static org.assertj.core.api.Assertions.assertThat;

public class SessionTest extends DomainTest {

    private SpyEventPublisher eventPublisher;

    private SessionId SESSION_ID = SessionId.generate();
    private UserId USER_ID = new UserId("user@mixit.fr");

    @Before
    public void setUp() throws Exception {
        eventPublisher = new SpyEventPublisher();
    }

    @Test
    public void GivenAnExistingSessionWhenLoggingOutThenUserDisconnectedIsRaised() {
        // Given
        Session session = sessionFor(new UserConnected(SESSION_ID, USER_ID, Instant.now()));

        // When
        session.logout(eventPublisher);

        // Then
        UserDisconnected expected = new UserDisconnected(SESSION_ID, USER_ID);
        assertThat(eventPublisher.publishedEvents).containsExactly(expected);
    }

    @Test
    public void GivenADisconnectedSessionWhenLoggingOutThenNoEventIsRaised() {
        // Given
        Session session = sessionFor(
                new UserConnected(SESSION_ID, USER_ID, Instant.now()),
                new UserDisconnected(SESSION_ID, USER_ID));

        // When
        session.logout(eventPublisher);

        // Then
        assertThat(eventPublisher.publishedEvents).isEmpty();
    }

    private Session sessionFor(Event... events) {
        return new Session(history(events));
    }
}
